File: src/editor/export-assets.ts
/**
 * **Export Assets**: some of a project's PSDs, on their own.
 *
 * The third exit from the menu, and the only one that hands back *artwork*
 * rather than a program. Export site is a game you can serve and Export project
 * is the whole project as a `.idlewild`; both are all-or-nothing and both are
 * only useful to something that can read them. What was missing is the pictures:
 * the sprite sheets a tileset was sliced into, so they can go into another
 * engine or a document, and the source PSDs, so a file drawn on an iPad can be
 * opened on a desktop.
 *
 * So the sheet asks two questions and nothing else. **Which files** — a list
 * with a checkbox each, everything ticked to begin with, because "all of them"
 * is the common answer and un-ticking three is less work than ticking twelve.
 * And **what of them** — the generated assets, the PSDs themselves, or both.
 *
 * It lists what is on disk rather than what the document places. A file whose
 * placement has been deleted is still a file somebody drew, and the one thing
 * this export exists to rescue is artwork — refusing to hand back the only copy
 * of it would be the wrong kind of tidiness. A file the pipeline has never run
 * over says so on its row and cannot contribute generated assets, since it has
 * none.
 */

import { save as saveFileDialog } from "@tauri-apps/plugin-dialog";
import { h } from "../lib/dom";
import { openSheet } from "../lib/sheet";
import { publish, type PsdSummary } from "../lib/ipc";
import * as log from "../lib/log";

/** What the archive should carry, as the segmented control offers it. */
type Wanted = "assets" | "psds" | "both";

const WANTED: Array<{ value: Wanted; label: string; hint: string }> = [
  {
    value: "assets",
    label: "Assets",
    hint:
      "What the pipeline made: each file's data.json and the sprites and " +
      "tiles beside it. This is what another engine can read.",
  },
  {
    value: "psds",
    label: "PSDs",
    hint:
      "The source files, as Photoshop would open them — layers, the anchor " +
      "mark and the grid the artwork was drawn over.",
  },
  {
    value: "both",
    label: "Both",
    hint: "The source files and what the pipeline made from them, side by side.",
  },
];

export function openExportAssets(projectId: string, projectName: string): void {
  const sheet = openSheet({
    title: "Export Assets",
    subtitle: "Pick the PSDs, and what of them",
    width: 620,
  });

  const stem = projectName.replace(/[^\w-]+/g, "-").toLowerCase() || "idlewild";
  const chosen = new Set<string>();
  let wanted: Wanted = "both";

  const list = h("div", { class: "sheet-list scroll psd-picker" });
  const tally = h("div", { class: "sheet-row-key m" });
  const exportButton = h("button", {
    class: "btn btn-primary",
    text: "Export",
    onClick: () => void run(),
  }) as HTMLButtonElement;

  const refresh = () => {
    const n = chosen.size;
    tally.textContent = `${n} of ${files.length} selected`;
    exportButton.disabled = n === 0;
  };

  let files: PsdSummary[] = [];

  const run = async () => {
    // Read before the sheet closes: closing it drops the checkboxes, and the
    // dialog that follows is a round trip through the OS.
    const keys = [...chosen];
    const want = wanted;
    sheet.close();
    try {
      const path = await saveFileDialog({
        defaultPath: `${stem}-assets.zip`,
        filters: [{ name: "Zip archive", extensions: ["zip"] }],
      });
      if (!path) return;
      await publish.assets(
        projectId,
        path,
        keys,
        want !== "psds",
        want !== "assets",
      );
      log.info(
        `Exported ${keys.length} ${keys.length === 1 ? "PSD" : "PSDs"} ` +
          `(${want === "both" ? "assets and sources" : want}) → ${path}`,
      );
    } catch (err) {
      log.error("Export Assets failed:", err);
    }
  };

  // Which files. Built before the segmented control below it, because it is
  // the answer somebody came here to give.
  sheet.body.append(
    h(
      "div",
      { class: "sheet-row" },
      h("div", { class: "sheet-row-key m", text: "Files" }),
      h(
        "div",
        { class: "sheet-row-value picker-actions" },
        h("button", {
          class: "panel-btn",
          text: "All",
          onClick: () => setAll(true),
        }),
        h("button", {
          class: "panel-btn",
          text: "None",
          onClick: () => setAll(false),
        }),
        tally,
      ),
    ),
    list,
    h(
      "div",
      { class: "sheet-row control" },
      h("div", { class: "sheet-row-key m", text: "Export" }),
      h("div", { class: "sheet-row-value" }, segmented()),
    ),
  );

  sheet.actions.append(
    exportButton,
    h("button", { class: "btn btn-ghost", text: "Cancel", onClick: sheet.close }),
  );

  const boxes = new Map<string, HTMLInputElement>();

  function setAll(on: boolean): void {
    for (const [key, box] of boxes) {
      box.checked = on;
      if (on) chosen.add(key);
      else chosen.delete(key);
    }
    refresh();
  }

  /**
   * The three answers to "what of them", as one control.
   *
   * A segmented row rather than two checkboxes: the three states are the ones
   * anybody wants, and two boxes would let somebody tick neither — an export
   * with nothing in it, refused at the far end of a save dialog.
   */
  function segmented(): HTMLElement {
    const hint = h("div", { class: "check-hint" });
    const seg = h("div", { class: "seg" });
    const buttons = WANTED.map((option) =>
      h("button", {
        class: "seg-opt",
        "aria-pressed": String(option.value === wanted),
        text: option.label,
        onClick: () => {
          wanted = option.value;
          for (const [i, button] of buttons.entries()) {
            button.setAttribute(
              "aria-pressed",
              String(WANTED[i].value === wanted),
            );
          }
          hint.textContent = option.hint;
        },
      }),
    );
    seg.append(...buttons);
    hint.textContent = WANTED.find((o) => o.value === wanted)?.hint ?? "";
    return h("div", { class: "seg-stack" }, seg, hint);
  }

  void load();

  async function load(): Promise<void> {
    try {
      files = await publish.listPsds(projectId);
    } catch (err) {
      log.error("Could not read this project's PSDs:", err);
    }

    if (files.length === 0) {
      list.appendChild(
        h("div", {
          class: "sheet-list-hint",
          text: "This project has no PSDs yet. Add an image, or draw one.",
        }),
      );
      refresh();
      return;
    }

    for (const file of files) {
      const box = h("input", {
        type: "checkbox",
        class: "check-box",
      }) as HTMLInputElement;
      box.checked = true;
      chosen.add(file.key);
      box.addEventListener("change", () => {
        if (box.checked) chosen.add(file.key);
        else chosen.delete(file.key);
        refresh();
      });
      boxes.set(file.key, box);

      list.appendChild(
        h(
          "label",
          { class: "psd-pick" },
          box,
          h("span", { class: "psd-pick-name", text: `${file.key}.psd` }),
          h("span", {
            class: "sheet-list-hint m",
            // Size, and the one thing that changes what an export of this row
            // can contain.
            text: file.hasAssets ? size(file.bytes) : `${size(file.bytes)} · not processed`,
          }),
        ),
      );
    }
    refresh();
  }
}

/** A file size a person reads, rather than a number of bytes. */
function size(bytes: number): string {
  if (bytes >= 1_000_000) return `${(bytes / 1_000_000).toFixed(1)} MB`;
  if (bytes >= 1_000) return `${Math.round(bytes / 1_000)} KB`;
  return `${bytes} B`;
}

File: src-tauri/src/export_assets.rs
//! **Export Assets**: some of a project's PSDs, on their own.
//!
//! Publish has always had two exits and neither of them is this one. *Export
//! site* is a runnable game and *Export project* is the whole project as a
//! `.idlewild`; both are all-or-nothing, and both hand back something that is
//! only useful to a program. What was missing is the artwork: the sprite sheets
//! a tileset was sliced into, so they can go into another engine, and the source
//! PSDs, so a file drawn on an iPad can be opened on a desktop.
//!
//! So this takes a **list of keys** and says, per key, whether to write the
//! processed output, the source file, or both. The shapes inside the archive are
//! the store's own — `psd/<key>.psd` and `assets/<key>/…` — because that is the
//! layout every other part of this app already describes, and a second one
//! invented for the zip would be a second thing to learn.
//!
//! A key the project has not got is skipped rather than failed on. The list
//! comes from a picker built from `list`, so the only way to ask for a missing
//! one is to have deleted it between opening the sheet and pressing the button —
//! and losing the other nine files to that is not a trade worth making. What
//! *is* an error is an archive with nothing in it, which would otherwise be a
//! zip somebody has to open to discover was empty.

use crate::{psd_pipeline, store};
use std::io::Write;
use std::path::Path;
use zip::write::SimpleFileOptions;

/// **Export Assets**: the command the menu item reaches.
///
/// Written straight to the path the save dialog gave, like the two exports
/// beside it. The dialog hands back a `file://` URL on iPadOS rather than a
/// path, which is what `source_path` is for.
#[tauri::command]
pub fn export_assets_zip(
    id: String,
    path: String,
    keys: Vec<String>,
    assets: bool,
    psds: bool,
) -> Result<(), String> {
    write_zip(
        &id,
        &keys,
        Wanted { assets, psds },
        &crate::psd_write::source_path(&path),
    )
}

/// Every PSD in a project, for the picker Export Assets opens with.
#[tauri::command]
pub fn list_project_psds(id: String) -> Result<Vec<PsdSummary>, String> {
    list(&id)
}

/// One of a project's PSDs, as the picker lists it.
#[derive(Debug, Clone, serde::Serialize)]
pub struct PsdSummary {
    /// The key, which is the filename without `.psd`.
    pub key: String,
    /// Size of the source file, for the "what am I about to export" line.
    pub bytes: u64,
    /// Whether the pipeline has run over it, so there is anything generated to
    /// export. A file written but never processed has none.
    #[serde(rename = "hasAssets")]
    pub has_assets: bool,
}

/// Every PSD in a project's `psd/`, by key, alphabetically.
///
/// Read off the directory rather than out of the document, and deliberately: a
/// file whose placement has been deleted is still a file somebody drew, and the
/// point of this export is to get artwork out. The document's own list would
/// quietly refuse to hand back the one thing that cannot be recovered any other
/// way.
pub fn list(project_id: &str) -> Result<Vec<PsdSummary>, String> {
    let dir = store::psd_dir(project_id)?;
    let mut out = Vec::new();
    for entry in std::fs::read_dir(&dir)
        .map_err(|e| e.to_string())?
        .flatten()
    {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("psd") {
            continue;
        }
        let Some(key) = path.file_stem().and_then(|s| s.to_str()) else {
            continue;
        };
        out.push(PsdSummary {
            key: key.to_string(),
            bytes: entry.metadata().map(|m| m.len()).unwrap_or(0),
            has_assets: psd_pipeline::is_processed(project_id, key),
        });
    }
    out.sort_by(|a, b| a.key.cmp(&b.key));
    Ok(out)
}

/// What to put in the archive.
#[derive(Debug, Clone, Copy)]
pub struct Wanted {
    /// The pipeline's output: `data.json` and the sprites and tiles beside it.
    pub assets: bool,
    /// The source `.psd` itself.
    pub psds: bool,
}

/// Build the zip and write it where the user asked for it.
///
/// In memory first, like `publish::write_zip` and for the same reason: the
/// archive writer wants a seekable sink, and the bytes go to disk from here
/// rather than back across the IPC boundary as base64.
pub fn write_zip(
    project_id: &str,
    keys: &[String],
    wanted: Wanted,
    dest: &Path,
) -> Result<(), String> {
    let bytes = build_zip(project_id, keys, wanted)?;
    if let Some(parent) = dest.parent() {
        std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    std::fs::write(dest, bytes).map_err(|e| format!("Cannot write {dest:?}: {e}"))
}

pub fn build_zip(project_id: &str, keys: &[String], wanted: Wanted) -> Result<Vec<u8>, String> {
    if keys.is_empty() {
        return Err("Nothing selected to export".into());
    }
    if !wanted.assets && !wanted.psds {
        return Err("Choose the generated assets, the source PSDs, or both".into());
    }

    let meta = store::read_meta(project_id)?;
    let root = crate::publish::sanitise_name(&meta.name);
    let mut buf = Vec::new();
    let mut written = 0usize;
    {
        let mut zip = zip::ZipWriter::new(std::io::Cursor::new(&mut buf));
        let options =
            SimpleFileOptions::default().compression_method(zip::CompressionMethod::Deflated);

        for key in keys {
            // A bad key is the picker's problem, not the archive's — and an
            // invalid one cannot reach a path from here either way.
            let Ok(source) = psd_pipeline::psd_path(project_id, key) else {
                continue;
            };
            if wanted.psds && source.is_file() {
                let bytes =
                    std::fs::read(&source).map_err(|e| format!("Cannot read {key}.psd: {e}"))?;
                zip.start_file(format!("{root}/psd/{key}.psd"), options)
                    .map_err(|e| e.to_string())?;
                zip.write_all(&bytes).map_err(|e| e.to_string())?;
                written += 1;
            }
            if wanted.assets {
                let Ok(dir) = psd_pipeline::output_dir(project_id, key) else {
                    continue;
                };
                if dir.is_dir() {
                    written += add_dir(&mut zip, &dir, &format!("{root}/assets/{key}/"), options)?;
                }
            }
        }

        if written == 0 {
            return Err(
                "None of those PSDs have anything to export — the files may have \
                 been renamed or deleted since this sheet was opened."
                    .into(),
            );
        }

        zip.finish().map_err(|e| e.to_string())?;
    }
    Ok(buf)
}

/// Copy a directory into the archive, and say how many files that was.
fn add_dir<W: Write + std::io::Seek>(
    zip: &mut zip::ZipWriter<W>,
    dir: &Path,
    prefix: &str,
    options: SimpleFileOptions,
) -> Result<usize, String> {
    let mut written = 0;
    let mut stack = vec![dir.to_path_buf()];
    while let Some(current) = stack.pop() {
        for entry in std::fs::read_dir(&current)
            .map_err(|e| e.to_string())?
            .flatten()
        {
            let path = entry.path();
            if path.is_dir() {
                stack.push(path);
                continue;
            }
            let rel = path
                .strip_prefix(dir)
                .unwrap_or(&path)
                .to_string_lossy()
                .replace('\\', "/");
            let bytes = std::fs::read(&path).map_err(|e| e.to_string())?;
            zip.start_file(format!("{prefix}{rel}"), options)
                .map_err(|e| e.to_string())?;
            zip.write_all(&bytes).map_err(|e| e.to_string())?;
            written += 1;
        }
    }
    Ok(written)
}
